Technical

Per-Morten opened the pull request:
Not to be merged yet, only exists to get feedback.

Per-Morten committed:
ref #35: Add introduction to overview

Per-Morten committed:
ref #35: Add requirements to overview

Per-Morten committed:
ref #35: Add measurements to overview

Per-Morten committed:
ref #35: Add technical to overview

Per-Morten committed:
ref #35: Add implementation to overview

Per-Morten committed:
ref #35: Add benchmarking to overview

Per-Morten committed:
ref #35: Add testing to overview

Per-Morten committed:
ref #35: Add discussion to overview

Per-Morten committed:
ref #35: Add future work to overview

Per-Morten committed:
ref #35: Add process to overview

Per-Morten committed:
ref #35: Add conclusion to overview

Per-Morten committed:
fix #35: Add overview_format and overview.pdf

Added the overview format, as well as a release folder where we can put
non pdf's that are actually finished, and not just stuff we are building
for ourselves.

Per-Morten committed:
ref #35: Expand ecs in title

Expanded the ECS to Entity Component System within the title of the
thesis.

Per-Morten committed:
fix #35: Move sections based on feedback

Moved some sections around based on feedback from Mariusz.
The Metrics was moved to closer to the benchmark chapter, as it is more
relevant there.
Also moved the section related to Entity Component Systems after we have
introduced Suttung and NOX, and also closer to our high level goal.

Per-Morten committed:
ref #63: Remove meetinglog

Removed the meeting log section, as we will have our own appendix for
that.

Per-Morten committed:
ref #63: Remove deployment

Removed the deployment section as that is not relevant for us.

Per-Morten committed:
ref #63: Add benchmark section

Per-Morten committed:
ref #63: Add conclusion

Per-Morten committed:
ref #63: Add discussion

Per-Morten committed:
ref #63: Add introduction

Per-Morten committed:
ref #63: Add measurements

Per-Morten committed:
ref #63: Add process

Per-Morten committed:
ref #63: Add requirements

Per-Morten committed:
ref #63: Add technical

Per-Morten committed:
ref #63: Add testing

Per-Morten committed:
ref #63: Add future work

Per-Morten committed:
ref #63: Add implementation

Per-Morten committed:
ref #63: Add preface with subimport

Per-Morten committed:
fix #63: Add appendix headings

Per-Morten committed:
ref #53: Add section on virtual functions

Started working on the section related to virtual functions, there is
still more to do on that section though, it is not finished yet.

Per-Morten committed:
ref #53: Add citations to virtual functions

Added some citations to the section on virtual functions, as well as
some extra text.
The segment should be close to complete, just need to clear up the last
two todos.

Per-Morten committed:
ref #53: Add section on multi-threading

Added a short section on multi-threading, don't really know what more to
write there.

Per-Morten committed:
ref #52: Add section on templates

Added a section on templates, don't know how much is needed here, should
clear it with Mariusz.

Per-Morten committed:
ref #55: Add section on mobile platform

Added section on mobile platform, unsure if what I wrote is actually
good enough.

Per-Morten committed:
ref #53: Add source for data layout

Per-Morten committed:
ref #56: Add section on familiarity

Added a short section on Suttungs desire for familiarity within the
requirements.

Per-Morten committed:
ref #57: Add section on safety

Added a small section on safety, don't know how much to write here.

Per-Morten committed:
ref #68: Add small section on intention

Added a small section on the intention behind the system.

Per-Morten committed:
ref #68: Add todo to intention

Added a small todo to the intention to help remember some potential
things to write about.

Per-Morten committed:
ref #59: Add section on lifecycle

Added a section on lifecycle, more is still probably needed though.

Per-Morten committed:
ref #69: Add section on update cycle

Added a small section quickly explaining the update cycle.
This will be further explained in the detailed architecture on the
EntityManager.

Per-Morten committed:
ref #60: Add small section on parallelism

Added a small WIP section on parallelism.

Per-Morten committed:
ref #68: Add todo on grouping in intention

Per-Morten committed:
ref #61: Add todo to event system

Per-Morten committed:
ref #52: Add todo on templates

Per-Morten committed:
ref #48: Add small section on game industry

Added a small introductory section on the game industry.

Per-Morten committed:
ref #65: Add section on organization

Added a section regarding the organization of the document,
more is probably needed here.

Per-Morten committed:
ref #70: Add section on operation types

Added a section on operation types, closing in on finishing it.

Per-Morten committed:
ref #72: Add section on meta information

Added a section on the meta information, think it is close to finished
as well.

Per-Morten committed:
ref #73: Add section on component collection

Added a section on the component collection, it should be close to
complete.

Per-Morten committed:
ref #35: Move high level architecture sections

Added some todos and comments, also removed the smart handle section, as
it is now covered within the component collection.

Per-Morten committed:
ref #74: Add section on component

Added a section on the base component class, and removed the
consequences section.

Per-Morten committed:
ref #72: Remove consequences section

Removed the consequences section of the meta information structure,
and added some proper labels and sectin references.

Per-Morten committed:
ref #70: Add link to operation types

Added a link to the operation types which can be used when discussing
longevity, using the is_detected rather than the homebrew SFINAE
solution.

Per-Morten committed:
ref #73: Remove consequences in collection

Removed the consequences section from the component collection.
Also added the entry for the citation within the collection.

Per-Morten committed:
ref #59: Add label to lifecycle

Per-Morten committed:
ref #71: Add todo in type identifier

Per-Morten committed:
Merge branch 'requirements' into technical

Merging the branches as there are no reasons to keep them separate,
it just gets in the way.

Per-Morten committed:
Merge branch 'introduction' into technical

Merging these branches as there are no reason to keep them separate,
it just gets in the way.

Per-Morten committed:
ref #1: Remove images

Removed images that were remaining from the template example.

Per-Morten committed:
ref #53: Add label in performance

Per-Morten committed:
ref #52: Add label in templates

Per-Morten committed:
ref #76: Add headline for longevity

Added a headline for longevity to help reminding us that this needs to
be written.

Per-Morten committed:
ref #1: Add needed custom commands

Added some custom commands that helps when writing the document.

Per-Morten committed:
ref #77: Start section on lock_free_allocator

Started writing the section on the lock_free_allocator.

lohne150 commented on thesis/inc/introduction/game_industry.tex:
```diff
@@ -0,0 +1,34 @@
+allowing developers to quickly combine different modules into new combinations creating new functionality,
+and content.
+
+\subsection{Re-usability}
```
No need for the - between re and usability. Reusability is a perfectly valid word by itself

lohne150 commented on thesis/inc/introduction/game_industry.tex:
```diff
@@ -0,0 +1,34 @@
+or what platforms the game is being developed for.
+
+\subsection{Game Engines}
+Modern video games are often created using game engines. A game engine can be seen as a foundation/framework
```
Im more fond of using or instead of / in this kind of case.

lohne150 commented on thesis/inc/introduction/game_industry.tex:
```diff
@@ -0,0 +1,34 @@
+
+\subsection{Game Engines}
+Modern video games are often created using game engines. A game engine can be seen as a foundation/framework
+containing a set of core modules needed for a game, like a rendering engine, or physics engine.
```
I would suggest replacing it with "like a rendering or physics module"

lohne150 commented on thesis/inc/introduction/game_industry.tex:
```diff
@@ -0,0 +1,34 @@
+Modern video games are often created using game engines. A game engine can be seen as a foundation/framework
+containing a set of core modules needed for a game, like a rendering engine, or physics engine.
+Different game engines have different strengths and weaknesses, depending on different factors,
+examples include what sort of game genre the engine was designed for.
```
could also talk about engines for different platforms, and for different level of programming competence. Would say that unity is much more beginner friendly than unreal.

lohne150 commented on thesis/inc/introduction/organization.tex:
```diff
@@ -0,0 +1,9 @@
+development issues, benchmarking, future work and a discussion on the longevity
+of some of the design choices taken in the development of the product.
+The second section will mainly cover the development process of the product.
+How did we plan our process, work methodology, tools used etc.
```
We have a detailed description of all the parts in the first section, but just say etc... for the second one. Do we not yet know all the parts of the second section?

Per-Morten replied:
Currently unknown.

lohne150 commented on thesis/inc/requirements/familiarity.tex:
```diff
@@ -0,0 +1,7 @@
+\section{Familiarity}
+Suttung uses their current NOX engine for most of their game development.
+They are quite familiar with their current actor system, and expressed a desire that the new
+ECS should not deviate to far from the original actor system in terms of programming paradigms
```
too far

lohne150 commented on thesis/inc/requirements/familiarity.tex:
```diff
@@ -0,0 +1,7 @@
+They are quite familiar with their current actor system, and expressed a desire that the new
+ECS should not deviate to far from the original actor system in terms of programming paradigms
+and familiarity.
+This would allow Suttung to continue creating games without to much loss of productivity while
```
too much

lohne150 commented on thesis/inc/requirements/performance.tex:
```diff
@@ -0,0 +1,59 @@
+\subsection{Virtual Functions}
+\label{subsec:requirements_performance_virtual_functions}
+\todo{Refer to Game Engine Architecture?}
+Virtual functions is C++'s way of supporting dynamic function binding,allowing for run-time polymorphism.
```
no space between comma and the word after "binding,allowing"

lohne150 commented on thesis/inc/requirements/performance.tex:
```diff
@@ -0,0 +1,59 @@
+however, many compilers choose the same tactic; creating a virtual table.\cite{wikipedia_virtual_method_table}
+\todo{Link to standard? (c++14 or c++17)?, Ask Mariusz if draft is ok, as the real thing costs money}
+
+A virtual table can be viewed as a data structure containing function pointers to virtual functions that should be called in a derived class.
```
Replace should with will. I'm quite confident that the virtual functions will be called in the derived classes. Could also restructure to just saying "... functions that is called in a derived class."

lohne150 commented on thesis/inc/technical/detailed_architecture/component.tex:
```diff
@@ -0,0 +1,43 @@
+\subsection{Component}
+\label{subsec:detailed_component}
+The component is a simple base class that all
```
Somewhat wierd sentence. Would rather write "The component is a simple base class that all components must inherit from". I feel everything i discarded in the new sentence was just "fluff" and did not add anything to the old sentence.

lohne150 commented on thesis/inc/technical/detailed_architecture/component.tex:
```diff
@@ -0,0 +1,43 @@
+that allows for the generic behavior without direct type information.
+
+\subsubsection{Motivation}
+As with a lot of the building block structures within the ECS the need for a base component
```
Need a comma or period in this sentence after ECS.

lohne150 commented on thesis/inc/technical/detailed_architecture/component.tex:
```diff
@@ -0,0 +1,43 @@
+This would mean that a component would potentially avoid unneeded members like the entity manager pointer.
+It would also lead to even less coupling within the systems,
+as the components would not need to inherit from the component base class.
+However going with this approach would require even more opaque type usage within the ECS,
```
Comma after however

lohne150 commented on thesis/inc/technical/detailed_architecture/component.tex:
```diff
@@ -0,0 +1,43 @@
+\subsubsection{Pros}
+The main advantage of the component base class over the alternatives of the component pair is that it allows
+for a use of the component base class as an interface.
+Additionally inheriting from a base class allows for declaration of all member functions that is used within the ECS,
```
Comma after additionally, or just removing the word entirely.

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+\label{subsec:detailed_component_collection}
+The component collection is where all the different building blocks
+comes together.
+As the name suggests the component collection is responsible for containing
```
Is the addition of "As the name suggests" really needed?

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+comes together.
+As the name suggests the component collection is responsible for containing
+a collection of a specific component type.
+The collection was design for fast contiguous processing of component groups internally within the ECS,
```
designed*

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+As the name suggests the component collection is responsible for containing
+a collection of a specific component type.
+The collection was design for fast contiguous processing of component groups internally within the ECS,
+while giving sufficient to external lookups on individual components.
```
while giving sufficient "missing word?" to external.
Seems like there is a word missing here.

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+this can be a costly operation, but is mitigated by using a simple growth strategy.
+When a collection needs to reallocate it will expand the buffer size based it's growth factor.
+The new size of the collection is simply its old size multiplied with its growth factor.
+This leads to an amortized constant complexity when adding a new component to the collection,
```
Is amortized the right word to use here? From what I can find via google, there was no definition fitting with how it is used here.

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+\paragraph{Index Mapping}
+Collections are often queried for their components, meaning that some form of lookup is needed.
+Several approaches could have been taken to solve this problem,
+the path taken in the collection was to add an indexing structure on top of the collection.
```
"Several approaches could have been taken to solve this problem, the path taken in the collection" seems a bit over the top, and not really necessary. Would replace it with "The problem was resolved by adding an indexing structure..."

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+allowing for faster lookup through a binary search.
+
+\paragraph{Iterator Invalidation}
+When using contiguous memory iterator invalidation is often a problem, and this case is no different.
```
"When using a contiguous memory iterators, invalidation is often a problem..."

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+
+\paragraph{Iterator Invalidation}
+When using contiguous memory iterator invalidation is often a problem, and this case is no different.
+After a reallocation any references to objects in the old memory location is invalidated,
```
"are invalidated". We are talking about references, so plural should be used here as well.

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+An alternative could be to not partition the components, however that would mean that another way of storing
+lifecycle information would be needed, for example a state variable.
+It would also lead to worse branch prediction, as components would need to be queried about their state.
+However avoiding partitioning would also lead to less iterator invalidation, as moving memory around would
```
Just remove the however

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+Requiring full contiguous memory might be a bit to extreme for the collection, as this leads to
+reallocation in several cases. Going for a semi contiguous approach akin to that of a standard C++ deque\footnote{\url{http://en.cppreference.com/w/cpp/container/deque}},
+or a linked list containing several contiguous memory segments.
+Such approach could lead to lower costs of reallocation, while still allowing for efficient use of caching mechanisms
```
Such an approach*

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+or a linked list containing several contiguous memory segments.
+Such approach could lead to lower costs of reallocation, while still allowing for efficient use of caching mechanisms
+when iterating.
+However searching, removal and partitioning could suffer from such an implementation.
```
Either remove however, or add a comma after it

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+A simpler solution could be to go for an external reference system like the one described by Reinalter\cite{molecular_matters_dod_external_references},
+where users are given an index and generation into a sparse array, which indexes into the regular component array.
+The main reason for not going with this approach was originally to not use extra space for the outer sparse arrays,
+however when the index map was introduced to allow for faster searching, that memory was needed anyway.
```
comma after however

lohne150 commented on thesis/inc/technical/detailed_architecture/component_collection.tex:
```diff
@@ -0,0 +1,135 @@
+
+\paragraph{Memory Movement}
+As mentioned earlier there is a lot of memory movement overhead required for this structure to avoid fragmentation and
+maintain the partitioning. This can be quite expensive when working with types that are heavy to move.
```
maintaining*

lohne150 commented on thesis/inc/technical/detailed_architecture/lock_free_allocator.tex:
```diff
@@ -0,0 +1,92 @@
+\subsubsection{Explanation}
+As mentioned earlier the allocator is a linear allocator,
+this means that while the allocator supports allocations of many different sizes,
+however it only supports mass deallocation, i.e. freeing all memory at once.
```
remove however

lohne150 commented on thesis/inc/technical/detailed_architecture/meta_information.tex:
```diff
@@ -0,0 +1,60 @@
+
+\subsubsection{Motivation}
+The meta information structure was mainly motivated by the requirement of avoiding heavy template and virtual functionality\footnote{Detailed in sections \ref{sec:requirements_templates} and \ref{subsec:requirements_performance_virtual_functions}}.
+NOX ECS would need a way to store general type information that could be use to write generic functionality, without templates,
```
Omit general, isn't "type information" enough? Also it should say "that could be used to write"

lohne150 commented on thesis/inc/technical/high_level_architecture/lifecycle.tex:
```diff
@@ -0,0 +1,64 @@
+
+\subsubsection{Hibernation}
+Hibernation is the first stage within the lifecycle where a component can be said to be valid.
+This stage is either directly after the constructor or after an \todo{a or an?} user requested initialize function has been executed on a component.
```
a

lohne150 commented on thesis/inc/technical/high_level_architecture/lifecycle.tex:
```diff
@@ -0,0 +1,64 @@
+
+\subsubsection{Active}
+A component that is active is one that is currently running within the system.
+This component has all it's resources available and is currently getting its update function called.
```
its*

Per-Morten committed:
ref #77: Elaborate section on lock_free_allocator

Elaborated a bit on the lock_free_allocator, it should be ready for
review now.

Per-Morten committed:
ref #1: Fix duplicate source

Per-Morten committed:
ref #1: Fix duplicate keyword

Per-Morten committed:
ref #48: Fix issues mentioned in review

Per-Morten committed:
ref #65: Fix issues mentioned in review

Per-Morten committed:
ref #56: Fix issues mentioned in review

Per-Morten committed:
Add todo's based on feedback

Per-Morten committed:
ref #55: Fix sources after period

Per-Morten committed:
ref #53: Fix issues mentioned in review

Per-Morten committed:
ref #52: Fix issues mentioned in review

Per-Morten committed:
Add todo's based on feedback

Per-Morten committed:
ref #74: Fix issues mentioned in review

Per-Morten committed:
ref #73: Fix issues mentioned in review

Per-Morten committed:
ref #81: Add label to EntityManager

Per-Morten committed:
ref #82: Add label to event

Per-Morten committed:
ref #77: Fix issues mentioned in review

Per-Morten committed:
ref #72: Fix issues mentioned in review

Per-Morten committed:
ref #70: Fix issues mentioned in review

Per-Morten committed:
ref #68: Fix issues mentioned in review

Per-Morten committed:
ref #59: Fix issues mentioned in review

Per-Morten committed:
ref #60: Fix issues mentioned in review

Per-Morten committed:
ref #69: Add comments from feedback

Per-Morten committed:
Add quick feedback comments from supervisor

Per-Morten committed:
fix #83: Add requirement keyword

Added a new keyword for adding and referencing requirements.

Per-Morten committed:
ref #56: Add requirement to familiarity

Per-Morten committed:
ref #55: Add requirement to mobile platforms

Per-Morten committed:
ref #53: Add requirements to performance

Per-Morten committed:
ref #57: Add requirements to safety

Per-Morten committed:
ref #52: Add requirements to templates

Per-Morten committed:
ref #74: Fix minor issues in component

Per-Morten committed:
ref #73: Fix minor issues in component collection

Per-Morten committed:
ref #72: Fix minor issues in meta information

Per-Morten committed:
ref #70: Fix minor issues in operation types

Per-Morten committed:
ref #73: Add clarification to collection title

Per-Morten committed:
ref #77: Add clarification to allocator heading

Per-Morten committed:
ref #72: Add clarification to meta information

Per-Morten committed:
ref #84: Add section on thread pool

Added a section the thread pool, think it should be what we need.

Per-Morten committed:
ref #79: Add label to execution layer

Added a label to the execution layer, allowing for refferences to it.

Per-Morten committed:
ref #69: Add label to update cycle

Per-Morten committed:
ref #81: Add section on entity manager

Added a section discussing the entity manager, think it should be close
to finished.

Commit:
ref #79: Add pseudocode for execution layers

Added pseudeocode for the initialization of data for the execution
layers algorithm. Displays how the data is collected and transformed
before it is finally used in the actual algorith.|

Commit:
ref #79: Add a semi complete of execution layers

Per-Morten committed:
ref #82: Add section on event

Added a section on event, not entirely finished, but it is a start.

Per-Morten committed:
ref #71: Add section on type identifier

Added a short section on type identifiers, don't know if more is really
needed.

Per-Morten committed:
ref #85: Add template for lock-free stack

Added the template for the lock-free stack, however it is not written
yet.

Per-Morten committed:
Merging

Minor merge to get the partially written execution layers section.

Per-Morten committed:
ref #85: Add section on lock_free_stack

Should now be closing in on done with the lock free stack.
Also added some labels to the lock_free_allocator, and a
task to discuss the ABA problem, which the allocator avoids,
by requiring the explicit clearing of the allocator.

ref #77

Commit:
ref #79: Add improvements to execution layers

FIlled in the remaining sections, and reworked a lot of the already
existing ones too.

Per-Morten committed:
ref #77: Add ABA problem to lock free allocator

Added a short paragraph about the ABA problem to the lock free allocator

Per-Morten committed:
ref #85: Add clarification to lock_free_stack

Added a clarification to the lock free stack, regarding the poping
semantics.
Also added a short paragraph on the use of third party software.

Per-Morten committed:
Merge branch 'issue_#79' into technical

Per-Morten committed:
ref #87: Add section on modules

Added a section on C++ modules to the longevity portion of the techincal
section. Not sure if more is needed.

Per-Morten committed:
ref #87: Add section on seq_cst architectures

Added a section for sequentially consistent architecture within the
techincal section.

Per-Morten committed:
ref #87: Add section on transactional memory

Added a section on transactional memory, unsure how much is really
needed.

Per-Morten committed:
ref #54: Add section on memory usage

Added a section on memory usage related to the requirements.

Per-Morten committed:
ref #53: Add section on contiguous memory

Added a section on contiguous memory in terms of performance in the
requirements.

Per-Morten committed:
ref #60: Elaborate on parallelism

Elaborated a bit on the parallelism section.

Per-Morten committed:
ref #59: Remove todo in lifecycle

Removed a todo in lifecycle after the relevant question was cleared with
suttung.

Per-Morten committed:
ref #61: Add section on event system

Added a small section on the event system. Not really sure what is
needed as most is covered in the detailed section.

Commit:
ref #79: Write more on test cases

Written a bit more on the test cases, more notably in the metrics
sections.
Talked about how the bash shell command time is used, and how the cycle
count is collected.

Per-Morten committed:
ref #51: Add introduction section on ecss

Added an introduction concentrating on entity component systems.
Also added a section on audiences.

Per-Morten committed:
Various fixes from feedback

Did various fixes on feedback gotten on the thesis.

Per-Morten committed:
ref #62: Add section on choices

Added a small section on choices that were taken when developing the NOX
ecs.

Per-Morten committed:
ref #52: Add section on high level objective

Added the section on high level objective, don't know if more is needed,
however a more detailed objective statement will be given within the
requirements.

Per-Morten committed:
ref #88: Add small section on roles

Added a small section on roles, don't know how much is needed here.

Per-Morten committed:
ref #89: Add section on tools

Added a small section on the tools used when developing. More is needed.

Per-Morten committed:
Merge branch 'issue_#79' into technical

Per-Morten committed:
ref #1: Remove finished todo

Per-Morten committed:
ref #90: Start self evaluation

Started working on the self evaluation section.

Per-Morten committed:
ref #91: Start criticism

Started writing the section on criticism.

Per-Morten committed:
ref #1: Move around on process

Moved around on the section within the process as suggested by reviewer.

Per-Morten committed:
ref #53: Fix requirement comment

Moved part of the requirement out into a comment.

Per-Morten committed:
ref #75: Add uml diagram

Added a small uml diagram to the high level architecture segment.

Per-Morten committed:
ref #59: Add lifecycle figure

Added a figure to the lifecycle section.

Per-Morten committed:
ref #73: Add partitioning picture to collection

Added a picture showing the partitioning to the component collection.

Per-Morten committed:
ref #73: Add index map figure to collection

Added a picture of the index map to the component collection.

Per-Morten committed:
ref #92: Add section on future work

More is still needed but some parts are now added to the future work
section.

Per-Morten committed:
Add title and description

Added a title to the project, and gaveit a short description, however it
should probably be rewritten a bit.

Per-Morten committed:
ref #93: Add preface

Added a suggested preface, more might be needed.

Per-Morten committed:
ref #67: Add extra information to lots of sources

Commit:
ref #79: Add mathtools package

Commit:
ref #79: Finish up identifying metrics

Finished up identifying metrics, but still needs a few small additions
that won't take too long to do.

Per-Morten committed:
Merge branch 'issue_#79' into technical

Per-Morten committed:
ref #94: Fix wording in audience

Commit:
ref #79: Fix various small issues

Commit:
ref #79: Write about compilation and fast spawning

Wrote about the compilation and fast spawning test cases.
The fast spawning is not yet complete, but is on its way.

Per-Morten committed:
ref #77: Fix missing atomic qualifier

Per-Morten committed:
ref #67: Add more elaborate sources

Per-Morten committed:
Merge branch 'issue_#79' into technical

Commit:
ref #79: Finish writing up test cases

Per-Morten committed:
Merge branch 'issue_#79' into technical

Per-Morten committed:
ref #1: Remove user testing section

Removed the user testing section in the thesis, as we never did any user
testing.

Per-Morten committed:
ref #73: Add section on search to collection

Added a section on the overhead of doing searches to the component
collection

Per-Morten committed:
ref #1: Remove todo

Per-Morten committed:
ref #79: Add optimization toggles notes

Added some notes on the optimization toggles to the test cases section.

Per-Morten committed:
ref #1: Minor cleanup in tools section

Per-Morten committed:
ref #95: Add sections to scrum process

Added some sections to the scrum process section.

Per-Morten committed:
Thesis update

Sorry for lack of commit, but we are way to tired to do this properly
now.

Per-Morten committed:
Wrote benchmarking section

Added a bunch of data on the benchmarking section.

Per-Morten committed:
Remove difficulties, it is not relevant anymore

Per-Morten committed:
Minor cleanup in benchmarking

Per-Morten committed:
Finish conclusion

Per-Morten committed:
Add discussion and criticism

Per-Morten committed:
Add future work section

Per-Morten committed:
Add criticism of process

Per-Morten committed:
Add section on plan

Per-Morten committed:
Add section on estimation

Per-Morten committed:
Add note to requirements

Per-Morten committed:
Remove implementation as we have the detailed architecture instead

Per-Morten committed:
Add label to execution layers

Per-Morten committed:
Add label to component collection

Per-Morten committed:
Add meta information example to operation types

Per-Morten committed:
Remove implementation from thesis
